tests/utils.py

```diff
@@ -265,19 +266,23 @@ def setup_test_homeserver(
         cur.close()
         db_conn.close()
 
+    def inject_cache(obj, kw: dict):
+        # Install @cache_in_self attributes
+        for key, val in kw.items():
+            setattr(obj, "_" + key, val)
+
+        obj.tls_server_context_factory = Mock()
+        obj.tls_client_options_factory = Mock()
+
```

Comment:
If you're going to inject cache-specific attributes, at least be explicit about it

also, deduplicated code for 2 cases down below where it needs to be injected just after creating the homeserver objects and *before* any branch logic